algorithms/src/std_algorithms/numeric/Kokkos_InclusiveScan.hpp

```diff
@@ -72,31 +71,13 @@ struct InclusiveScanDefaultFunctor {
         m_first_dest(std::move(first_dest)) {}
 
   KOKKOS_FUNCTION
-  void operator()(const IndexType i, value_type& update,
+  void operator()(const IndexType i, ValueType& update,
                   const bool final_pass) const {
-    const auto tmp = value_type{m_first_from[i], false};
-    this->join(update, tmp);
+    update += m_first_from[i];
 
     if (final_pass) {
-      m_first_dest[i] = update.val;
-    }
-  }
-
-  KOKKOS_FUNCTION
-  void init(value_type& update) const {
-    update.val        = {};
-    update.is_initial = true;
-  }
-
-  KOKKOS_FUNCTION
-  void join(volatile value_type& update,
-            volatile const value_type& input) const {
-    if (update.is_initial) {
-      update.val = input.val;
-    } else {
-      update.val = update.val + input.val;
+      m_first_dest[i] = update;
     }
-    update.is_initial = false;
   }
 };
```

Comment:
why has this changed? this won't work for a value_type that does not have an identity. 
The previous impl was done general enough to handle any type. If we want to make this more efficient I think that we have to specialize the impl for when we have a type that has a reduction/scan identity, similarly to how you did the exclusive scan.